boutvecma/encoder.py

```diff
@@ -1,7 +1,47 @@
+import os
+from copy import deepcopy
 from easyvvuq.encoders import BaseEncoder
+from boutdata.data import BoutOptions, BoutOptionsFile
+
 
 class BOUTEncoder(BaseEncoder, encoder_name="bout++_encoder"):
+    def __init__(self, template_input=None):
+        """Read an existing BOUT.inp file to use as a template.
+        If no input is given, an empty set of options will be created
+
+        Example
+        -------
+
+        from boutvecma.encoder import BOUTEncoder
+        encoder = BOUTEncoder("data/BOUT.inp")
+        """
+        if template_input:
+            self._options = BoutOptionsFile(template_input)
+        else:
+            self._options = BoutOptions()
+
+    def encode(self, params={}, target_dir=""):
+        """Create a BOUT.inp file in target_dir with modified parameters.
+
+        Sub-sections are specified using colons in the params keys
+
+        Example
+        -------
+
+        encoder.encode({"section:key":42}, target_dir="data")
+
+        modifies key so that the file "data/BOUT.inp" contains
+
+        [section]
+        key = 42
+
+        """
+        options = deepcopy(self._options)
 
-    def encode(self, params={}, target_dir=''):
-        pass
+        for key, value in params.items():
+            options[key] = value
 
+        # Note: Here options could be BoutOptions or BoutOptionsFile
+        # so can't just use options.write
+        with open(os.path.join(target_dir, "BOUT.inp"), "w") as f:
```

Comment:
Am I right in thinking that we also need to deal with copying any other files/data into the target directory here? I know we don't usually have anything else in a BOUT++ data dir but it's possible we might want to copy in grid files etc. (if we're trying to run lots of jobs in parallel there could be issues trying to access netcdf grid files from different processes at once?)

Reply:
Yes, good point. This is probably easy to do if it's just a file with the key `gridfile`, but might be harder to do in general?

Reply:
Maybe we call something like `self.additional_setup` at the end of this method and register an no-op form here to allow the user to extend?